modules/axfs/src/fs/ext4fs.rs

```
@@ -0,0 +1,408 @@
+            .map_err(|e| <i32 as TryInto<AxError>>::try_into(e).unwrap())?;
+
+        let t = file.file_truncate(size);
+
+        let _ = file.file_close();
+        t.map(|_v| ()).map_err(|e| e.try_into().unwrap())
+    }
+
+    fn rename(&self, src_path: &str, dst_path: &str) -> VfsResult {
+        debug!("rename from {} to {}", src_path, dst_path);
+        let mut file = self.0.lock();
+        file.file_rename(src_path, dst_path)
+            .map(|_v| ())
+            .map_err(|e| e.try_into().unwrap())
+    }
+
+    fn as_any(&self) -> &dyn core::any::Any {
+        self as &dyn core::any::Any
+    }
+}
+
+impl Drop for FileWrapper {
+    fn drop(&mut self) {
+        let mut file = self.0.lock();
+        debug!("Drop struct FileWrapper {:?}", file.get_path());
+        file.file_close().expect("failed to close fd");
+        drop(file); // todo
+    }
+}
+
+impl KernelDevOp for Disk {
+    //type DevType = Box<Disk>;
+    type DevType = Disk;
+
+    fn read(dev: &mut Disk, mut buf: &mut [u8]) -> Result<usize, i32> {
+        debug!("READ block device buf={}", buf.len());
+        let mut read_len = 0;
+        while !buf.is_empty() {
+            match dev.read_one(buf) {
+                Ok(0) => break,
+                Ok(n) => {
+                    let tmp = buf;
+                    buf = &mut tmp[n..];
+                    read_len += n;
+                }
+                Err(_e) => return Err(-1),
+            }
+        }
+        debug!("READ rt len={}", read_len);
+        Ok(read_len)
+    }
+
+    fn write(dev: &mut Self::DevType, mut buf: &[u8]) -> Result<usize, i32> {
+        debug!("WRITE block device buf={}", buf.len());
+        let mut write_len = 0;
+        while !buf.is_empty() {
+            match dev.write_one(buf) {
+                Ok(0) => break,
+                Ok(n) => {
+                    buf = &buf[n..];
+                    write_len += n;
+                }
+                Err(_e) => return Err(-1),
```

Here you can return a canonical error message instead of a simple -1?

---

replace LinuxError::EIO with -1